code/datums/components/tippable.dm

```diff
@@ -0,0 +1,187 @@
+		datum/callback/pre_tipped_callback,
+		datum/callback/post_tipped_callback,
+		datum/callback/post_untipped_callback)
+
+	if(!isliving(parent))
+		return COMPONENT_INCOMPATIBLE
+
+	src.tip_time = tip_time
+	src.untip_time = untip_time
+	src.self_right_time = self_right_time
+	src.pre_tipped_callback = pre_tipped_callback
+	src.post_tipped_callback = post_tipped_callback
+	src.post_untipped_callback = post_untipped_callback
+
+/datum/component/tippable/RegisterWithParent()
+	RegisterSignal(parent, COMSIG_ATOM_ATTACK_HAND, .proc/interact_with_tippable)
+
+/datum/component/tippable/UnregisterFromParent()
+	UnregisterSignal(parent, COMSIG_ATOM_ATTACK_HAND)
+
+/datum/component/tippable/Destroy()
+	if(pre_tipped_callback)
+		QDEL_NULL(pre_tipped_callback)
+	if(post_tipped_callback)
+		QDEL_NULL(post_tipped_callback)
+	if(post_untipped_callback)
+		QDEL_NULL(post_untipped_callback)
+	return ..()
+
+/*
+ * Attempt to interact with [source], either tipping it or helping it up.
+ *
+ * source - the mob being tipped over
+ * user - the mob interacting with source
+ * modifiers - list of on click modifiers (we only tip mobs over using right click!)
+ */
+/datum/component/tippable/proc/interact_with_tippable(mob/living/source, mob/user, modifiers)
+	SIGNAL_HANDLER
+
+	var/mob/living/living_user = user
+	if(DOING_INTERACTION_WITH_TARGET(user, source))
+		return
+	if(istype(living_user) && !living_user.combat_mode)
+		return
+
+	if(is_tipped)
+		INVOKE_ASYNC(src, .proc/try_untip, source, user)
+	else if(LAZYACCESS(modifiers, RIGHT_CLICK))
+		INVOKE_ASYNC(src, .proc/try_tip, source, user)
+
+	return COMPONENT_CANCEL_ATTACK_CHAIN
+
+/*
+ * Try to tip over [tipped_mob].
+ * If the mob is dead, or optional callback returns a value, or our do-after fails, we don't tip the mob.
+ * Otherwise, upon completing of the do_after, tip over the mob.
+ *
+ * tipped_mob - the mob being tipped over
+ * tipper - the mob tipping the tipped_mob
+ */
+/datum/component/tippable/proc/try_tip(mob/living/tipped_mob, mob/tipper)
+	if(tipped_mob.stat)
+		return
+
+	if(pre_tipped_callback?.Invoke(tipper))
+		return
+
+	if(tip_time > 0)
+		to_chat(tipper, span_warning("You begin tipping over [tipped_mob]..."))
+		to_chat(tipped_mob, span_userdanger("[tipper] begins tipping you over!"))
+		tipper.visible_message(span_warning("[tipper] begins tipping over [tipped_mob]."), ignored_mobs = list(tipped_mob, tipper))
+		if(!do_after(tipper, tip_time, target = tipped_mob))
+			to_chat(tipper, span_danger("You fail to tip over [tipped_mob]."))
+			return
+	do_tip(tipped_mob, tipper)
+
+/*
+ * Actually tip over the mob, setting it to tipped.
+ * Also invoking any callbacks we have, with the tipper as the argument,
+ * and set a timer to right our self-right our tipped mob if we can.
+ *
+ * tipped_mob - the mob who was tipped
+ * tipper - the mob who tipped the tipped_mob
+ */
+/datum/component/tippable/proc/do_tip(mob/living/tipped_mob, mob/tipper)
+	if(QDELETED(tipped_mob))
+		return
```

Comment:
```suggestion
		CRASH("do_tip() called for QDELETED tipped_mob")
```
This should never happen, given the prior checks you added.
Whenever you implement sanity checks, make sure to give them a nice debugging message, instead of silently failing.